File: opcua-webservice-node/src/server/addressspace/reactiveobject/UaObjectTransactionManager.ts
import { Attributes, StatusCodes } from "opcua-webapi";
import { DataTypeIds, makeUaStatusCode, UaBrowseDescription, UaDataValue, UaNodeId, UaNodeIdType } from "opcua-webapi-ts";
import {
    CallContext,
    HistoryReadContext,
    ReadContext,
    ServiceContext,
    UaBrowseAdditionalInfo,
    UaChildId,
    UaChildIdentifier,
    UaInstanceIdentifier,
    UaObjectId,
    UaObjectIdentifier,
    WriteContext,
    WriteVariableValue,
} from "../../types";
import {
    UaBrowseNodeTransaction,
    UaBrowseTransaction,
    UaHistoryReadTransaction,
    UaMethodCallTransaction,
    UaReadNodeTransaction,
    UaReadTransaction,
    UaWriteTransaction,
} from "../../service/transactions/base";
import { 
    UaBrowseMemberTransaction ,
    UaBrowseObjectTransaction,
    UaCallMethodTransaction,
    UaReadDataHistoryTransaction,
    UaReadEventHistoryTransaction,
    UaReadMemberAttributeTransaction,
    UaReadObjectAttributeTransaction,
    UaReadVariableValueTransaction,
    UaWriteNodeTransaction,
    UaWriteObjectAttributeTransaction,
    UaWriteVariableValueTransaction } from "../../service/transactions/reactiveobject";
import { NodeManagerReactiveObject } from "../nodemanager/NodeManagerReactiveObject";

type GroupedObjectHandles = {
    objectId: UaObjectIdentifier;
    handleIds: Array<number>;
};

export class UaObjectTransactionManager {
    private readonly nodeManager: NodeManagerReactiveObject;

    constructor(nodeManager: NodeManagerReactiveObject) {
        this.nodeManager = nodeManager;
    }

    getBrowseTransaction(
        context: ServiceContext,
        nodeToBrowse: UaBrowseDescription,
        additionalInfo: UaBrowseAdditionalInfo,
        handleId: number,
    ): UaBrowseTransaction {
        const transactionNothingToDo = new UaBrowseTransaction(
            context,
            nodeToBrowse,
            additionalInfo,
            handleId,
        );

        if (!this.isReactiveObjectNodeId(nodeToBrowse.nodeId)) {
            return new UaBrowseNodeTransaction(
                context,
                nodeToBrowse,
                additionalInfo,
                handleId,
                this.nodeManager,
            );
        }

        const identifier = this.getInstanceIdentifier(nodeToBrowse.nodeId);
        if (identifier === null) {
            transactionNothingToDo.setStatusCode(makeUaStatusCode(StatusCodes.BadNodeIdInvalid));
            return transactionNothingToDo;
        }

        const objectType = this.nodeManager.findObjectType(identifier.objectId);
        if (objectType === null) {
            transactionNothingToDo.setStatusCode(makeUaStatusCode(StatusCodes.BadNodeIdUnknown));
            return transactionNothingToDo;
        }

        const instanceDeclaration = this.nodeManager.findInstanceDeclaration(identifier.objectId);
        const objectId = new UaObjectId(identifier.objectId.id, instanceDeclaration);

        if (identifier.childId === null) {
            return new UaBrowseObjectTransaction(
                context,
                nodeToBrowse,
                additionalInfo,
                handleId,
                objectType,
                objectId,
                this.nodeManager,
            );
        }

        return new UaBrowseMemberTransaction(
            context,
            nodeToBrowse,
            additionalInfo,
            handleId,
            objectType,
            objectId,
            identifier.childId,
            this.nodeManager,
        );
    }

    getReadTransactions(context: ReadContext, handleIds: Array<number>): Array<UaReadTransaction> {
        const handleIdsForNodes: Array<number> = [];
        const handleIdsForObjectAttributes = new Map<number, UaInstanceIdentifier>();
        const handleIdsForObjectMemberAttributes = new Map<number, UaInstanceIdentifier>();
        const handleIdsForObjectVariableValues = new Map<number, UaInstanceIdentifier>();

        for (const handleId of handleIds) {
            const nodeToRead = context.nodesToRead[handleId];
 
            if (!this.isReactiveObjectNodeId(nodeToRead.nodeId)) {
                handleIdsForNodes.push(handleId);
                continue;
            }

            const identifier = this.getInstanceIdentifier(nodeToRead.nodeId);
            const objectType = (identifier === null) ? null : this.nodeManager.findObjectType(identifier.objectId);

            if (identifier === null || objectType === null) {
                handleIdsForNodes.push(handleId);
                continue;
            }

            if (nodeToRead.attributeId === Attributes.Value) {
                if (identifier.childId === null) {
                    handleIdsForNodes.push(handleId);
                } else if (!identifier.childId.methodNode) {
                    handleIdsForObjectVariableValues.set(handleId, identifier);
                } else {
                    handleIdsForObjectMemberAttributes.set(handleId, identifier);
                }

                continue;
            }

            if (identifier.childId !== null) {
                handleIdsForObjectMemberAttributes.set(handleId, identifier);
            } else {
                handleIdsForObjectAttributes.set(handleId, identifier);
            }
        }

        const transactions: Array<UaReadTransaction> = [];

        if (handleIdsForNodes.length > 0) {
            transactions.push(new UaReadNodeTransaction(context, handleIdsForNodes, this.nodeManager));
        }

        let groupsForObjectAttributes = this.groupByObject(handleIdsForObjectAttributes);
        for (const group of groupsForObjectAttributes) {
            transactions.push(new UaReadObjectAttributeTransaction(
                context,
                group.handleIds,
                group.objectId,
                this.nodeManager,
            ));
        }

        let groupsForObjectMemberAttributes = this.groupByObject(handleIdsForObjectMemberAttributes);
        for (const group of groupsForObjectMemberAttributes) {
            const handleIdsByChildId = new Map<string, { childId: UaChildIdentifier; handleIds: Array<number> }>();

            for (const handleId of group.handleIds) {
                const memberId = handleIdsForObjectMemberAttributes.get(handleId)?.childId;
                if (memberId === null || memberId === undefined) {
                    continue;
                }

                const key = memberId.toString();
                const currentGroup = handleIdsByChildId.get(key);
                if (currentGroup === undefined) {
                    handleIdsByChildId.set(key, { childId: memberId, handleIds: [handleId] });
                    continue;
                }

                currentGroup.handleIds.push(handleId);
            }

            for (const childGroup of handleIdsByChildId.values()) {
                transactions.push(new UaReadMemberAttributeTransaction(
                    context,
                    group.objectId,
                    childGroup.childId,
                    childGroup.handleIds,
                    this.nodeManager,
                ));
            }
        }

        let groupsForObjectVariableValues = this.groupByObject(handleIdsForObjectVariableValues);
        for (const group of groupsForObjectVariableValues) {
            const handleIdsAndVariableIds = new Map<number, UaChildId>();

            for (const handleId of group.handleIds) {
                const memberId = handleIdsForObjectVariableValues.get(handleId)?.childId;
                if (memberId === null || memberId === undefined) {
                    continue;
                }

                handleIdsAndVariableIds.set(handleId, new UaChildId(memberId.path, memberId.pathL2));
            }

            transactions.push(new UaReadVariableValueTransaction(
                context,
                group.objectId,
                handleIdsAndVariableIds,
                this.nodeManager,
            ));
        }

        return transactions;
    }

    getWriteTransactions(context: WriteContext, handleIds: Array<number>): Array<UaWriteTransaction> {
        const handleIdsForNodes: Array<number> = [];
        const handleIdsForObjectAttributes = new Map<number, UaInstanceIdentifier>();
        const handleIdsForWriteVariableValues = new Map<number, UaInstanceIdentifier>();

        for (const handleId of handleIds) {
            const nodeToWrite = context.nodesToWrite[handleId];
 
            if (!this.isReactiveObjectNodeId(nodeToWrite.nodeId)) {
                handleIdsForNodes.push(handleId);
                continue;
            }

            const identifier = this.getInstanceIdentifier(nodeToWrite.nodeId);
            const objectType = identifier === null ? null : this.nodeManager.findObjectType(identifier.objectId);

            if (identifier === null || objectType === null) {
                handleIdsForNodes.push(handleId);
                continue;
            }

            if (nodeToWrite.attributeId === Attributes.Value) {
                if (identifier.childId !== null) {
                    handleIdsForWriteVariableValues.set(handleId, identifier);
                } else {
                    handleIdsForNodes.push(handleId);
                }

                continue;
            }

            if (identifier.childId === null && identifier.objectId.instanceDeclId === null) {
                handleIdsForObjectAttributes.set(handleId, identifier);
            } else {
                handleIdsForNodes.push(handleId);
            }
        }

        const transactions: Array<UaWriteTransaction> = [];

        if (handleIdsForNodes.length > 0) {
            transactions.push(new UaWriteNodeTransaction(context, handleIdsForNodes));
        }

        let groupsForObjectAttributes = this.groupByObject(handleIdsForObjectAttributes);
        for (const group of groupsForObjectAttributes) {
            transactions.push(new UaWriteObjectAttributeTransaction(
                context,
                group.handleIds,
                group.objectId,
                this.nodeManager,
            ));
        }

        let groupsForWriteVariableValues = this.groupByObject(handleIdsForWriteVariableValues);
        for (const group of groupsForWriteVariableValues) {
            const handleIdsAndVariableValues = new Map<number, WriteVariableValue>();

            for (const handleId of group.handleIds) {
                const memberId = handleIdsForWriteVariableValues.get(handleId)?.childId;
                const value = context.nodesToWrite[handleId]?.value;
                if (memberId === null || memberId === undefined || value === undefined) {
                    continue;
                }

                handleIdsAndVariableValues.set(
                    handleId,
                    new WriteVariableValue(
                        new UaChildId(memberId.path, memberId.pathL2),
                        new UaDataValue(value),
                    ),
                );
            }

            transactions.push(new UaWriteVariableValueTransaction(
                context,
                group.objectId,
                handleIdsAndVariableValues,
                this.nodeManager,
            ));
        }

        return transactions;
    }

    getMethodCallTransaction(context: CallContext, handleId: number): UaMethodCallTransaction {
        const callRequest = context.methodsToCall[handleId];
        const objectIdentifier = this.getInstanceIdentifier(callRequest.objectId);
        const methodIdentifier = this.getInstanceIdentifier(callRequest.methodId);

        if (objectIdentifier === null || methodIdentifier === null) {
            return new UaMethodCallTransaction(context, handleId);
        }

        return new UaCallMethodTransaction(
            context,
            handleId,
            objectIdentifier,
            methodIdentifier,
            this.nodeManager,
        );
    }

    getHistoryReadTransaction(context: HistoryReadContext, handleId: number): UaHistoryReadTransaction {
        const transactionNothingToDo = new UaHistoryReadTransaction(context, handleId);
        const nodeToRead = context.nodesToRead[handleId];
        if (nodeToRead === undefined || !this.isReactiveObjectNodeId(nodeToRead.nodeId)) {
            return transactionNothingToDo;
        }

        const identifier = this.getInstanceIdentifier(nodeToRead.nodeId);
        if (identifier === null) {
            transactionNothingToDo.setStatusCode(makeUaStatusCode(StatusCodes.BadNodeIdUnknown));
            return transactionNothingToDo;
        }

        const objectType = this.nodeManager.findObjectType(identifier.objectId);
        if (objectType === null) {
            transactionNothingToDo.setStatusCode(makeUaStatusCode(StatusCodes.BadNodeIdUnknown));
            return transactionNothingToDo;
        }

        if (identifier.childId === null) {
            if (context.historyReadDetails.typeId.equal(UaNodeId.from(DataTypeIds.ReadEventDetails))) {
                return new UaReadEventHistoryTransaction(context, handleId, identifier, this.nodeManager);
            }
        } else if (
            context.historyReadDetails.typeId.equal(UaNodeId.from(DataTypeIds.ReadRawModifiedDetails)) ||
            context.historyReadDetails.typeId.equal(UaNodeId.from(DataTypeIds.ReadAtTimeDetails)) ||
            context.historyReadDetails.typeId.equal(UaNodeId.from(DataTypeIds.ReadProcessedDetails))
        ) {
            return new UaReadDataHistoryTransaction(context, handleId, identifier, this.nodeManager);
        }

        transactionNothingToDo.setStatusCode(makeUaStatusCode(StatusCodes.BadHistoryOperationUnsupported));
        return transactionNothingToDo;
    }

    private isReactiveObjectNodeId(nodeId: UaNodeId): boolean {
        return nodeId.identifierType === UaNodeIdType.BYTESTRING;
    }

    private getInstanceIdentifier(nodeId: UaNodeId): UaInstanceIdentifier | null {
        if (!this.isReactiveObjectNodeId(nodeId)) {
            return null;
        }

        const value = nodeId.value;
        return typeof value === "string" ? UaInstanceIdentifier.fromByteString(value) : null;
    }

    private groupByObject(handleIdsAndIdentifiers: Map<number, UaInstanceIdentifier>): Array<GroupedObjectHandles> {
        const groups = new Map<string, GroupedObjectHandles>();

        for (const [handleId, identifier] of handleIdsAndIdentifiers.entries()) {
            const key = identifier.objectId.toString();
            const group = groups.get(key);

            if (group === undefined) {
                groups.set(key, {
                    objectId: identifier.objectId,
                    handleIds: [handleId],
                });
                continue;
            }

            group.handleIds.push(handleId);
        }

        return [...groups.values()];
    }
}

File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/addressspace/reactiveobject/UaObjectTransactionManager.java
package org.opcfoundation.webserver.addressspace.reactiveobject;

import org.eclipse.milo.opcua.stack.core.AttributeId;
import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.types.builtin.ByteString;
import org.eclipse.milo.opcua.stack.core.types.builtin.DataValue;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.eclipse.milo.opcua.stack.core.types.builtin.StatusCode;
import org.eclipse.milo.opcua.stack.core.types.enumerated.IdType;
import org.eclipse.milo.opcua.stack.core.types.structured.*;
import org.opcfoundation.webapi.service.types.*;
import org.opcfoundation.webserver.addressspace.nodemanager.NodeManagerReactiveObject;
import org.opcfoundation.webserver.addressspace.nodes.UaObject;
import org.opcfoundation.webserver.service.transactions.base.*;
import org.opcfoundation.webserver.service.transactions.reactiveobject.*;
import org.opcfoundation.webserver.types.common.*;

import java.util.*;

public class UaObjectTransactionManager {
    private final NodeManagerReactiveObject nodeManager;

    public UaObjectTransactionManager(NodeManagerReactiveObject nodeManager)
    {
        this.nodeManager = nodeManager;
    }

    public UaBrowseTransaction getBrowseTransaction(
            ServiceContext context,
            BrowseDescription nodeToBrowse,
            UaBrowseAdditionalInfo additionalInfo,
            int handleId)
    {
        UaBrowseTransaction transactionNothingToDo = new UaBrowseTransaction(
                context,
                nodeToBrowse,
                additionalInfo,
                handleId);

        if (IdType.Opaque != nodeToBrowse.getNodeId().getType())
            return new UaBrowseNodeTransaction(
                    context,
                    nodeToBrowse,
                    additionalInfo,
                    handleId,
                    nodeManager);

        final UaInstanceIdentifier identifier = UaInstanceIdentifier.fromByteString((ByteString) nodeToBrowse.getNodeId().getIdentifier());
        if (null == identifier)
        {
            transactionNothingToDo.setStatusCode(StatusCode.of(StatusCodes.Bad_NodeIdInvalid));
            return transactionNothingToDo;
        }

        // Browse object
        UaReactiveObjectType objectType = nodeManager.findObjectType(identifier.getObjectId());

        if (null == objectType)
        {
            transactionNothingToDo.setStatusCode(StatusCode.of(StatusCodes.Bad_NodeIdUnknown));
            return transactionNothingToDo;
        }

        UaObject instanceDeclaration = nodeManager.findInstanceDeclaration(identifier.getObjectId());
        UaObjectId objectId = new UaObjectId(identifier.getObjectId().getId(), instanceDeclaration);

        //System.out.println("NodeToBrowse: " + identifier);
        if (null == identifier.getChildId())
        {
            return new UaBrowseObjectTransaction(
                    context,
                    nodeToBrowse,
                    additionalInfo,
                    handleId,
                    objectType,
                    objectId,
                    nodeManager);
        } else {
            return new UaBrowseMemberTransaction(
                    context,
                    nodeToBrowse,
                    additionalInfo,
                    handleId,
                    objectType,
                    objectId,
                    identifier.getChildId(),
                    nodeManager);
        }
    }

    public List<UaReadTransaction> getReadTransactions(
            ReadContext context,
            List<Integer> handleIds)
    {
        List<Integer> handleIdsForNodes = new ArrayList<>();
        Map<Integer, UaInstanceIdentifier> handleIdsForObjectAttributes = new TreeMap<>();
        Map<Integer, UaInstanceIdentifier> handleIdsForObjectMemberAttributes = new TreeMap<>();
        Map<Integer, UaInstanceIdentifier> handleIdsForObjectVariableValues = new TreeMap<>();

        for (Integer handleId: handleIds)
        {
            ReadValueId nodeToRead = context.getNodesToRead().get(handleId);
            NodeId nodeId = nodeToRead.getNodeId();

            if (nodeId.getType() == IdType.Opaque)
            {
                UaInstanceIdentifier identifier = UaInstanceIdentifier.fromByteString((ByteString) nodeId.getIdentifier());
                UaReactiveObjectType objectType = null;
                if (null != identifier) objectType = nodeManager.findObjectType(identifier.getObjectId());

                if (null != identifier && null != objectType)
                {
                    if (nodeToRead.getAttributeId().intValue() == AttributeId.Value.id())
                    {
                        if (null != identifier.getChildId())
                        {
                            if (null == identifier.getChildId().getMethodNode() ||
                                    !identifier.getChildId().getMethodNode())
                            {
                                handleIdsForObjectVariableValues.put(handleId, identifier);
                                //System.out.println("Read value: " + identifier);
                            } else {
                                handleIdsForObjectMemberAttributes.put(handleId, identifier);
                                //System.out.println("Read arguments: " + identifier);
                            }
                        } else {
                            handleIdsForNodes.add(handleId);
                            //System.out.println("Read error, id invalid");
                        }
                    } else {
                        if (null != identifier.getChildId())
                        {
                            handleIdsForObjectMemberAttributes.put(handleId, identifier);
                            //System.out.println("Read object member attribute: " + identifier);
                        } else {
                            handleIdsForObjectAttributes.put(handleId, identifier);
                            //System.out.println("Read object attribute: " + identifier);
                        }
                    }
                } else {
                    handleIdsForNodes.add(handleId);
                    //System.out.println("Read error, id invalid");
                }
            } else {
                handleIdsForNodes.add(handleId);
                //System.out.println("Read: " + nodeId.toParseableString());
            }
        }

        List<UaReadTransaction> transactions = new ArrayList<>();

        if (!handleIdsForNodes.isEmpty())
        {
            UaReadNodeTransaction transaction = new UaReadNodeTransaction(
                    context,
                    handleIdsForNodes,
                    nodeManager);

            transactions.add(transaction);
        }

        if (!handleIdsForObjectAttributes.isEmpty())
        {
            Map<UaObjectIdentifier, List<Integer>> handleIdsByObjectId = new HashMap<>();

            for (Map.Entry<Integer, UaInstanceIdentifier> item: handleIdsForObjectAttributes.entrySet())
            {
                UaInstanceIdentifier identifier = item.getValue();
                handleIdsByObjectId.computeIfAbsent(identifier.getObjectId(), k-> new ArrayList<>()).add(item.getKey());
            }

            for (Map.Entry<UaObjectIdentifier, List<Integer>> item: handleIdsByObjectId.entrySet())
            {
                UaReadObjectAttributeTransaction transaction = new UaReadObjectAttributeTransaction(
                        context,
                        item.getValue(),
                        item.getKey(),
                        nodeManager);

                transactions.add(transaction);
            }
        }

        if (!handleIdsForObjectMemberAttributes.isEmpty())
        {
            Map<UaObjectIdentifier, List<Integer>> handleIdsByObjectId = new HashMap<>();

            for (Map.Entry<Integer, UaInstanceIdentifier> item: handleIdsForObjectMemberAttributes.entrySet())
            {
                UaInstanceIdentifier identifier = item.getValue();
                handleIdsByObjectId.computeIfAbsent(identifier.getObjectId(), k-> new ArrayList<>()).add(item.getKey());
            }

            for (Map.Entry<UaObjectIdentifier, List<Integer>> item: handleIdsByObjectId.entrySet())
            {
                Map<UaChildIdentifier, List<Integer>> handleIdsByChildId = new HashMap<>();

                for (Integer handleId: item.getValue())
                {
                    UaChildIdentifier memberId = handleIdsForObjectMemberAttributes.get(handleId).getChildId();
                    if (null == memberId) continue;

                    handleIdsByChildId.computeIfAbsent(memberId, k-> new ArrayList<>()).add(handleId);
                }

                for (Map.Entry<UaChildIdentifier, List<Integer>> child: handleIdsByChildId.entrySet())
                {
                    UaReadMemberAttributeTransaction transaction = new UaReadMemberAttributeTransaction(
                            context,
                            item.getKey(),
                            child.getKey(),
                            child.getValue(),
                            nodeManager);

                    transactions.add(transaction);
                }
            }
        }

        if (!handleIdsForObjectVariableValues.isEmpty())
        {
            Map<UaObjectIdentifier, List<Integer>> handleIdsByObjectId = new HashMap<>();

            for (Map.Entry<Integer, UaInstanceIdentifier> item: handleIdsForObjectVariableValues.entrySet())
            {
                UaInstanceIdentifier identifier = item.getValue();
                handleIdsByObjectId.computeIfAbsent(identifier.getObjectId(), k-> new ArrayList<>()).add(item.getKey());
            }

            for (Map.Entry<UaObjectIdentifier, List<Integer>> item: handleIdsByObjectId.entrySet())
            {
                Map<Integer, UaChildId> handleIdsAndVariableIds = new HashMap<>();

                for (Integer handleId: item.getValue())
                {
                    UaChildIdentifier memberId = handleIdsForObjectVariableValues.get(handleId).getChildId();
                    if (null == memberId) continue;
                    handleIdsAndVariableIds.put(handleId,new UaChildId(memberId.getPath(), memberId.getPathL2()));
                }

                UaReadVariableValueTransaction transaction = new UaReadVariableValueTransaction(
                        context,
                        item.getKey(),
                        handleIdsAndVariableIds,
                        nodeManager);

                transactions.add(transaction);
            }
        }

        return transactions;
    }

    public List<UaWriteTransaction> getWriteTransactions(
            WriteContext context,
            List<Integer> handleIds)
    {
        List<Integer> handleIdsForNodes = new ArrayList<>();
        Map<Integer, UaInstanceIdentifier> handleIdsForObjectAttributes = new TreeMap<>();
        Map<Integer, UaInstanceIdentifier> handleIdsForWriteVariableValues = new TreeMap<>();

        for (Integer handleId: handleIds)
        {
            WriteValue nodeToWrite = context.getNodesToWrite().get(handleId);
            NodeId nodeId = nodeToWrite.getNodeId();

            if (nodeId.getType() == IdType.Opaque)
            {
                UaInstanceIdentifier identifier = UaInstanceIdentifier.fromByteString((ByteString) nodeId.getIdentifier());
                UaReactiveObjectType objectType = null;
                if (null != identifier) objectType = nodeManager.findObjectType(identifier.getObjectId());

                if (null != identifier && null != objectType)
                {
                    if (nodeToWrite.getAttributeId().intValue() == AttributeId.Value.id())
                    {
                        if (null != identifier.getChildId())
                        {
                            handleIdsForWriteVariableValues.put(handleId, identifier);
                            //System.out.println("Write value: " + identifier);
                        } else {
                            handleIdsForNodes.add(handleId);
                            //System.out.println("Write not writable: " + identifier);
                        }
                    } else {
                        if (null == identifier.getChildId() &&
                                null == identifier.getObjectId().getInstanceDeclId())
                        {
                            handleIdsForObjectAttributes.put(handleId, identifier);
                            //System.out.println("Write object attribute: " + identifier);
                        } else {
                            handleIdsForNodes.add(handleId);
                            //System.out.println("Write not writable: " + identifier);
                        }
                    }
                } else {
                    handleIdsForNodes.add(handleId);
                    //System.out.println("Write not writable: " + identifier);
                }
            } else {
                handleIdsForNodes.add(handleId);
                //System.out.println("Write not writable: " + nodeId.toParseableString());
            }
        }

        List<UaWriteTransaction> transactions = new ArrayList<>();

        if (!handleIdsForNodes.isEmpty())
        {
            UaWriteNodeTransaction transaction = new UaWriteNodeTransaction(
                    context,
                    handleIdsForNodes);

            transactions.add(transaction);
        }

        if (!handleIdsForObjectAttributes.isEmpty())
        {
            Map<UaObjectIdentifier, List<Integer>> handleIdsByObjectId = new HashMap<>();

            for (Map.Entry<Integer, UaInstanceIdentifier> item: handleIdsForObjectAttributes.entrySet())
            {
                UaInstanceIdentifier identifier = item.getValue();
                handleIdsByObjectId.computeIfAbsent(identifier.getObjectId(), k-> new ArrayList<>()).add(item.getKey());
            }

            for (Map.Entry<UaObjectIdentifier, List<Integer>> item: handleIdsByObjectId.entrySet())
            {
                UaWriteObjectAttributeTransaction transaction = new UaWriteObjectAttributeTransaction(
                        context,
                        item.getValue(),
                        item.getKey(),
                        nodeManager);

                transactions.add(transaction);
            }
        }

        if (!handleIdsForWriteVariableValues.isEmpty())
        {
            Map<UaObjectIdentifier, List<Integer>> handleIdsByObjectId = new HashMap<>();

            for (Map.Entry<Integer, UaInstanceIdentifier> item: handleIdsForWriteVariableValues.entrySet())
            {
                UaInstanceIdentifier identifier = item.getValue();
                handleIdsByObjectId.computeIfAbsent(identifier.getObjectId(), k-> new ArrayList<>()).add(item.getKey());
            }

            for (Map.Entry<UaObjectIdentifier, List<Integer>> item: handleIdsByObjectId.entrySet())
            {
                Map<Integer, WriteVariableValue> handleIdsAndVariableValues = new HashMap<>();

                for (Integer handleId: item.getValue())
                {
                    UaChildIdentifier memberId = handleIdsForWriteVariableValues.get(handleId).getChildId();
                    DataValue value = context.getNodesToWrite().get(handleId).getValue();
                    if (null == memberId) continue;

                    handleIdsAndVariableValues.put(
                            handleId,
                            new WriteVariableValue(
                                    new UaChildId(memberId.getPath(), memberId.getPathL2()),
                                    value));
                }

                UaWriteVariableValueTransaction transaction = new UaWriteVariableValueTransaction(
                        context,
                        item.getKey(),
                        handleIdsAndVariableValues,
                        nodeManager);

                transactions.add(transaction);
            }
        }

        return transactions;
    }

    public UaMethodCallTransaction getMethodCallTransaction(
            CallContext context,
            int handleId)
    {
        CallMethodRequest callRequest = context.getMethodsToCall().get(handleId);

        UaInstanceIdentifier objectIdentifier = null;
        UaInstanceIdentifier methodIdentifier = null;

        if (callRequest.getObjectId().getType() == IdType.Opaque)
        {
            objectIdentifier = UaInstanceIdentifier.fromByteString((ByteString) callRequest.getObjectId().getIdentifier());
        }

        if (callRequest.getMethodId().getType() == IdType.Opaque)
        {
            methodIdentifier = UaInstanceIdentifier.fromByteString((ByteString) callRequest.getMethodId().getIdentifier());
        }

        if (null == objectIdentifier || null == methodIdentifier)
        {
            return new UaMethodCallTransaction(context, handleId);
        }

        return new UaCallMethodTransaction(context, handleId, objectIdentifier, methodIdentifier, nodeManager);
    }

    public UaHistoryReadTransaction getHistoryReadTransaction(
            HistoryReadContext context,
            int handleId)
    {
        UaHistoryReadTransaction transactionNothingToDo = new UaHistoryReadTransaction(
                context,
                handleId);

        HistoryReadValueId nodeToRead = context.getNodesToRead().get(handleId);
        if (IdType.Opaque != nodeToRead.getNodeId().getType()) return transactionNothingToDo;

        final UaInstanceIdentifier identifier = UaInstanceIdentifier.fromByteString((ByteString) nodeToRead.getNodeId().getIdentifier());
        if (null == identifier)
        {
            transactionNothingToDo.setStatusCode(StatusCode.of(StatusCodes.Bad_NodeIdUnknown));
            return transactionNothingToDo;
        }

        // Object type
        UaReactiveObjectType objectType = nodeManager.findObjectType(identifier.getObjectId());

        if (null == objectType)
        {
            transactionNothingToDo.setStatusCode(StatusCode.of(StatusCodes.Bad_NodeIdUnknown));
            return transactionNothingToDo;
        }

        if (null == identifier.getChildId())
        {
            if (context.getHistoryReadDetails() instanceof ReadEventDetails)
            {
                return new UaReadEventHistoryTransaction(context, handleId, identifier, nodeManager);
            }
        } else {
            if (context.getHistoryReadDetails() instanceof ReadRawModifiedDetails ||
                    context.getHistoryReadDetails() instanceof ReadAtTimeDetails ||
                    context.getHistoryReadDetails() instanceof ReadProcessedDetails)
            {
                return new UaReadDataHistoryTransaction(context,handleId,identifier, nodeManager);
            }
        }

        transactionNothingToDo.setStatusCode(StatusCode.of(StatusCodes.Bad_HistoryOperationUnsupported));
        return transactionNothingToDo;
    }
}
